core/domain/commit_commands_domain.py

```diff
@@ -0,0 +1,317 @@
+        error_msg = error_msg + 'The following extra keys are present: %s' % (
+            (',').join(extra_keys))
+
+    if missing_keys + extra_keys:
+        raise utils.ValidationError(error_msg)
+
+
+class BaseCommitCmd(object):
+    """Domain object for commit commands of a storage model.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
+
+    def __init__(self, name, parameters):
+        """Constructs a CommitCmd domain object.
+
+        Args:
+            name: str. The command name.
+            parameters: dict. The command parameters.
+        """
+        self.name = name
+        self.parameters = parameters
+        self.command_list = []
+
+    def validate(self):
+        """Checks that all parameters of a command are valid for the given
+        command.
+        """
+
+        required_keys = []
+        optional_keys = []
+        for cmd in self.command_list:
+            if cmd['name'] == self.name:
+                required_keys = cmd['required_keys']
+                optional_keys = cmd['optional_keys']
+                break
+        validate_dict_keys(self.parameters.keys(), required_keys, optional_keys)
+
+
+class ExplorationCommitCmd(BaseCommitCmd):
+    """Domain object for commit commands of a exploration.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
+
+    def __init__(self, name, parameters):
+        super(ExplorationCommitCmd, self).__init__(name, parameters)
+        self.command_list = [{
+            'name': exp_domain.CMD_CREATE_NEW,
+            'required_keys': ['category', 'title'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_ADD_STATE,
+            'required_keys': ['state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_DELETE_STATE,
+            'required_keys': ['state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_RENAME_STATE,
+            'required_keys': ['new_state_name', 'old_state_name'],
+            'optional_keys': []
+        }, {
+            'name': exp_domain.CMD_EDIT_STATE_PROPERTY,
+            'required_keys': ['property_name', 'state_name', 'new_value'],
+            'optional_keys': ['old_value']
+        }, {
+            'name': exp_domain.CMD_EDIT_EXPLORATION_PROPERTY,
+            'required_keys': ['property_name', 'new_value'],
+            'optional_keys': ['old_value']
+        }, {
+            'name': exp_domain.CMD_MIGRATE_STATES_SCHEMA_TO_LATEST_VERSION,
+            'required_keys': ['from_version', 'to_version'],
+            'optional_keys': []
+        }]
+
+
+class ExplorationRightsCommitCmd(BaseCommitCmd):
+    """Domain object for commit commands of a exploration rights.
+
+    Attributes:
+        name: str. The command name.
+        parameters: dict. The command parameters.
+    """
```

Comment:
There is no commit command defined for `exploration rights` in exp_domain. So, I left this class as empty. Same is the case with other empty classes. Should we have an empty init here as well?

Also, should the validation be made as required on init i.e we make a call to validate in the init itself. So, that whenever a command is created, it is validated as well?

Reply:
I think we need to introduce the commit cmds for rights objects in the same way as it is done for Exploration objects. You can see that there is stuff like CMD_CHANGE_PRIVATE_VIEWABILITY but it is all inlined as dicts and not represented as domain objects. (See rights_manager.py.) We should probably follow the format in exp_domain.py for this.

Reply:
Also, I think yeah, we should validate on init in this case. The initialization here is non-standard because the keys are not set in stone, so we need to make sure that the domain object we are building actually makes sense (based on the cmd type).

Reply:
Ok, Thanks for letting me know this. I will make the updates.

Reply:
Done.